enhancements/authentication/synchronizing-user-groups-on-login.md

```diff
@@ -0,0 +1,238 @@
+---
+title: synchronizing-user-groups-on-login
+authors:
+  - "@stlaz"
+reviewers:
+  - "@sur"
+  - "@slaskawi"
+approvers:
+  - "@deads2k"
+  - "@sttts"
+creation-date: 2021-05-22
+last-updated: yyyy-mm-dd
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Synchronizing User Groups on Login
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+OpenShift should be capable of synchronizing user groups from a 3rd party
+identity provider upon a user logging in to their OpenShift cluster.
+
+## Motivation
+
+In the real world, OpenShift is not the owner of user identities but these
+are stored in a 3rd party identity provider system. When OpenShift retrieves
+such an identity, it usually only gathers information about the user's unique
+identification in that IdP and optionally a username that's preferred for the
+user if it's different from its identifier.
+
+It is a common practice to group users in logical groups on the identity provider
+side and sometimes it would make sense to bring the groups over to the OpenShift
+cluster as well. The identity providers oftentimes allow a protocol-specific path
+to retrieve group information, which is something OpenShift should leverage in those
+cases where it is desired.
+
+### Goals
+
+- allow retrieving user's groups from a 3rd party OIDC identity provider into
+  OpenShift when such a user logs in
+
+### Non-Goals
+
+- implement group synchronization for an identity provider type other than
+  OIDC
+- keep the groups retrieved by OpenShift from a 3rd party identity provider
+  synchronized between user logins
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+I am an administrator of an OpenShift cluster that uses identities from
+my company's identity provider. The users of the identity provider are
+organized in a complex and well-though-out group structure. I would like
+OpenShift to respect the group structure so that I can assign privileges
+to the users on OpenShift side according to their group membership in the
+identity provider.
+
+### Implementation Details
+
+#### OIDC Group Retrieval
+
+OIDC core specification does not mention a standard `id_token` claim that
+would serve for user group storage and so none of the standard scopes define
+accessing it. That means retrieving the groups will very likely always require
+the user to configure `extraScopes` in their identity provider configuration
```

Comment:
OpenID Connect requires adding an additional scope (`openid`) to the token requests. Once we add it, we'll receive an ID Token, which should contain information about the user. As I mentioned earlier, there is no guarantee it will contain groups, but emitting them is up to the IdP's configuration. 

With that in mind, perhaps we could treat the Extra Scopes part as something optional? Most of the administrators should be able to make their IdPs to include groups in the token responses.

Reply:
> Once we add it, we'll receive an ID Token, which should contain information about the user. 

not all the information is going to be included, but perhaps we can get rid of most of this paragraph

`ExtraScopes` are already optional